theon/src/allocator.rs

```diff
@@ -47,36 +73,397 @@ pub(crate) struct BumpAlloc<T: Heap> {
 }
 
 impl<T: Heap> BumpAlloc<T> {
```

Comment:
Thanks for the nudge here. I really didn't like the `Heap` attribute, but couldn't figure out a better way when I introduced it back in 0f572bbfdb4cb1a23bc23bfae4fc360397bb82cf. Some background may be useful here, if for nothing else other than posterity. I originally had everything in terms of a slice, but changed it to use the `Heap` trait because I ran into problems with taking mut refs to mut statics, which, as per rust-lang/rust#114447, is turning into a hard error as of Rust 2024.

There are two distinct use cases that `BumpAlloc` handles:

1.  Serving as an arena for a generic allocator implementing the GlobalAlloc trait (and Allocator, but that came later). Initially that was just a dumb bump allocator with an empty `dealloc` method, but I started getting concerned about blowing up the heap with e.g., repeated resizing due to pushing into a `Vec` in a loop and things like that, hence implementing QuickFit. In this case, QuickFit conceptually rests on top of a bump allocator over a contiguous region of memory, but that region has no special requirements vis position or alignment (though aligned on some power of two boundary, such as 4KiB, might be nice just "because"). Because of the static mut issue, it was convenient to just have BumpAlloc own that region.
2. Serving as an arena allocator for physical memory regions. In these use cases, the actual place and alignment of the arena really do matter (I'm placing things like page tables for binaries in those regions when I load them, and where they are in memory is critical for hitless update). So while the allocator conceptually "owns" the region, I can't just move the arena into it as I did in case (1).

Hence the introduction of `Heap`, which let me keep the bulk of BumpAlloc's guts in terms of what amounted to the slice interface, while I papered over the difference between a moved static region of memory, and the pointer/length region pair.

But with the discussion in the upstream issue (in particular the comments around https://github.com/rust-lang/rust/issues/114447#issuecomment-1974891474) leads me to believe that rewriting `BumpAlloc` in terms of `Block`, and creating a Block from a `static mut` array that is only visible within a very limited scope, is probably safe. So I've gone ahead and done that, and I think the result is much better.